aleph-client/src/contract/mod.rs

```diff
@@ -0,0 +1,144 @@
+    pub fn contract_read(&self, conn: &Connection, message: &str, args: &[&str]) -> Result<Value> {
+        let payload = self.encode(message, args)?;
+        let request = self.contract_read_request(&payload);
+        let response = conn
+            .get_request(request)?
+            .context("RPC request error - there may be more info in node logs.")?;
+        let response_data = from_str::<serde_json::Value>(&response)?;
+        let hex_data = response_data["result"]["Ok"]["data"]
+            .as_str()
+            .context("Contract response data not found - the contract address might be invalid.")?;
+        self.decode_response(message, hex_data)
+    }
+
+    /// Executes a contract call.
+    pub fn contract_exec(
+        &self,
+        conn: SignedConnection,
+        message: &str,
+        args: &[&str],
+    ) -> Result<()> {
+        let data = self.encode(message, args)?;
+        let xt = compose_extrinsic!(
+            conn.as_connection(),
+            "Contracts",
+            "call",
+            GenericAddress::Id(self.address.clone()),
+            Compact(0u64),
+            Compact(Self::MAX_GAS),
+            None::<Option<u128>>,
+            data
+        );
+
+        try_send_xt(&conn, xt, Some("Contracts call"), XtStatus::InBlock)
+            .context("Failed to exec contract message")?;
+        Ok(())
+    }
+
+    fn contract_read_request(&self, payload: &[u8]) -> serde_json::Value {
+        let payload = hex::encode(payload);
+        json!({
+            "jsonrpc": "2.0",
+            "method": "contracts_call",
+            "params": [{
+                "origin": self.address,
+                "dest": self.address,
+                "value": 0,
+                "gasLimit": Self::MAX_READ_GAS,
+                "inputData": payload
+            }],
+            "id": 1
+        })
+    }
+
+    fn encode(&self, message: &str, args: &[&str]) -> Result<Vec<u8>> {
+        ContractMessageTranscoder::new(&self.ink_project).encode(message, args)
+    }
+
+    fn decode_response(&self, from: &str, contract_response: &str) -> Result<Value> {
+        let contract_response = contract_response.replacen("0x", "", 1);
+        let bytes = hex::decode(contract_response)?;
+        ContractMessageTranscoder::new(&self.ink_project).decode_return(from, &mut bytes.as_slice())
+    }
+}
+
+impl Debug for ContractInstance {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ContractInstance")
+            .field("address", &self.address)
+            .field("ink_project", &self.ink_project)
+            .finish()
+    }
+}
+
+/// Helper for loading contract metadata from a file.
+///
+/// The contract-metadata lib contains a similar function starting with version 0.2. It seems that
+/// version conflicts with some of our other dependencies, however, if we upgrade in the future we
+/// can drop this function in favour of their implementation.
```

Comment:
we don't have any policy about such situations, though it's not the first time we encounter such problem; I would loosely propose to add a short comment in `Cargo.toml`, above the mentioned dependency

Reply:
no, it's not a strong request; resolving